### PR TITLE
feat: create endpoint to get total products on admin dashboard

### PR DESCRIPTION
**Pull request by Leona-cloud**
## Description
Develop an API endpoint that retrieves the total number of products for the current month and compares it with the previous month's total. The response includes the total number of products and the percentage change from the previous month.

## Acceptance Criteria
The endpoint /api/v1/products/total is accessible via a GET request.
When the request is successful, the endpoint returns a 200 OK status.
If an unexpected server error occurs, the endpoint returns a 500 Internal Server Error status.
The response body of a 500 error contains an error message indicating that an unexpected error occurred.
Ensure that all unit and integration tests for the fetch total products endpoint are comprehensive and verify the functionality correctly.

## Purpose
This provides a backend service that allows fetching the total number of products for the current month and the percentage change from the previous month.

## Requirements
- [x] Implement an endpoint to view the total number of products for the current month.
- [x]  Return detailed information on the total number of products and the percentage change from the previous month.
- [x]  Handle situations where there are no products for the current or previous month.

## Expected Outcome
An API endpoint that returns the total number of products for the current month and the percentage change from the previous month.

## Endpoint URL: GET /api/v1/products/total
Responses
      Success Response
       - Code: 200 OK
       - Content:
       
```json
{
      "status": "string",
       "status_code": 200,
       "message": "string",
         "data": {
       "total_products": 0,
       "percentage_change": "+100.00% from last month"
       }
    }
```
image

Authentication Error:

When the current user is not properly authenticated to change the organization detail:

```json
{
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status": 401
}
```
image

  Error Responses
        - Code: 500 INTERNAL SERVER ERROR
         - Content:
```json
  {
  "error": "An unexpected error occurred"
  }
```

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-84/[feat]-get-total-products)

### CODE REVIEW

**Review by Heba-WebDev — APPROVED** [review]

File: src/modules/products/products.controller.ts [hunk 1]
```diff
@@ -1,60 +1,69 @@
...
 import { ProductsService } from './products.service';
 import { UpdateProductDTO } from './dto/update-product.dto';
+import { GetTotalProductsResponseDto } from './dto/get-total-products.dto';
+import { SuperAdminGuard } from 'src/guards/super-admin.guard';
 
+@ApiBearerAuth()
 @ApiTags('Products')
-@Controller('/organizations/:id/products')
+@Controller()
 export class ProductsController {
```
Comment (CFCIfe) [inline src/modules/products/products.controller.ts:13]:
```suggestion
@Controller('organizations')
```

File: src/modules/products/products.controller.ts
Comment (CFCIfe) [inline src/modules/products/products.controller.ts:1]: Revert the Controller to:

```suggestion
@Controller('organizations')
```

Then lets re-arrange the routes. All static routes e.g: `/products/total` should come top before parameterized routes e.g `/:orgId/products/:productId`

For order sake, a route like `/:orgId/products/total` must come before `/:orgId/products/:productId`.

File: src/modules/products/products.service.ts [hunk 2]
```diff
@@ -200,4 +201,53 @@ export class ProductsService {
...
+      const change = ((totalProductsThisMonth - totalProductsLastMonth) / totalProductsLastMonth) * 100;
+      percentageChange = `${change > 0 ? '+' : ''}${change.toFixed(2)}% from last month`;
+    } else if (totalProductsThisMonth > 0) {
+      percentageChange = `+100.00% from last month`;
+    }
+
+    return {
+      status: 'success',
+      status_code: HttpStatus.OK,
+      message: 'Total Products fetched successfully',
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:246]: Use SYS_MSG constant

File: src/modules/products/products.service.ts [hunk 3]
```diff
@@ -200,4 +201,53 @@ export class ProductsService {
...
+
+    const totalProductsLastMonth = await this.getTotalProductsForDateRange(lastMonthStarts, lastMonthEnds);
+
+    let percentageChange = '0% from last month';
+    if (totalProductsLastMonth !== 0) {
+      const change = ((totalProductsThisMonth - totalProductsLastMonth) / totalProductsLastMonth) * 100;
+      percentageChange = `${change > 0 ? '+' : ''}${change.toFixed(2)}% from last month`;
+    } else if (totalProductsThisMonth > 0) {
+      percentageChange = `+100.00% from last month`;
+    }
```
Comment (CFCIfe) [inline src/modules/products/products.service.ts:241]:
```suggestion
    if (totalProductsLastMonth !== 0) {
      const change = ((totalProductsThisMonth - totalProductsLastMonth) / totalProductsLastMonth) * 100;
      percentageChange = `${change > 0 ? '+' : ''}${change.toFixed(2)}% from last month`;
    } else if (totalProductsThisMonth > 0) {
      percentageChange = `+100.00% from last month`;
    } else {
        percentageChange = `No change from last month`;
    }
```

File: src/modules/products/products.controller.ts
Comment (CFCIfe) [inline src/modules/products/products.controller.ts:1]: Revert the Controller to:

```suggestion
@Controller('organizations')
```

Then lets re-arrange the routes. All static routes e.g: `/products/total` should come top before parameterized routes e.g `/:orgId/products/:productId`

For order sake, a route like `/:orgId/products/total` must come before `/:orgId/products/:productId`.

File: src/modules/products/products.controller.ts [hunk 4]
```diff
@@ -1,60 +1,69 @@
 import { Body, Controller, Delete, Get, HttpCode, Param, Patch, Post, UseGuards, Query } from '@nestjs/common';
-import { ApiBearerAuth, ApiBody, ApiOperation, ApiParam, ApiResponse, ApiTags } from '@nestjs/swagger';
+import { ApiBearerAuth, ApiBody, ApiOkResponse, ApiOperation, ApiParam, ApiResponse, ApiTags } from '@nestjs/swagger';
 import { OwnershipGuard } from '../../guards/authorization.guard';
 import { CreateProductRequestDto } from './dto/create-product.dto';
 import { ProductsService } from './products.service';
 import { UpdateProductDTO } from './dto/update-product.dto';
+import { GetTotalProductsResponseDto } from './dto/get-total-products.dto';
+import { SuperAdminGuard } from 'src/guards/super-admin.guard';
```
Comment (CFCIfe) [inline src/modules/products/products.controller.ts:8]:
```suggestion
import { SuperAdminGuard } from '../../guards/super-admin.guard';
```

File: src/modules/products/products.service.ts (hunk 2 above)
Comment (CFCIfe) [inline src/modules/products/products.service.ts:246]: Use system message constants

File: src/modules/products/products.service.ts (hunk 3 above)
Comment (CFCIfe) [inline src/modules/products/products.service.ts:241]: Add an else statement for when there is no product last month and this month

```suggestion
percentageChange = `No change from last month`;
```

**Review by CFCIfe — COMMENTED** [review]
Review comments and revert

**Review by davidolanrewaju — APPROVED** [review]

**Review by CFCIfe — APPROVED** [review]
LGTM.

**Pull request merged**